web/modules/custom/joinup_federation/src/Plugin/pipeline/Step/ThreeWayMerge.php

```diff
@@ -0,0 +1,269 @@
+    $incoming_ids = $this->getSparqlQuery()
+      ->graphs(['staging'])
+      ->execute();
+    /** @var \Drupal\rdf_entity\RdfInterface[] $incoming_entities */
+    $incoming_entities = $incoming_ids ? $this->getRdfStorage()->loadMultiple($incoming_ids, ['staging']) : [];
+
+    // Get the incoming entities that are stored also locally.
+    $local_ids = $this->getSparqlQuery()
+      ->graphs(['default', 'draft'])
+      ->condition('id', array_values($incoming_ids), 'IN')
+      ->execute();
+    /** @var \Drupal\rdf_entity\RdfInterface[] $local_entities */
+    $local_entities = $local_ids ? Rdf::loadMultiple($local_ids) : [];
+
+    /** @var \Drupal\rdf_entity\RdfInterface $incoming_entity */
+    foreach ($incoming_entities as $id => $incoming_entity) {
+      $bundle = $incoming_entity->bundle();
+
+      // The entity already exists.
+      if (isset($local_entities[$id])) {
+        $local_entity = $local_entities[$id];
+
+        // Check for bundle mismatch between the local and the incoming entity.
+        if ($local_entity->bundle() !== $bundle) {
+          $arguments = [
+            '%id' => $id,
+            '%incoming' => $incoming_entity->get('rid')->entity->getSingularLabel(),
+            '%local' => $local_entity->get('rid')->entity->getSingularLabel(),
+          ];
+          return [
+            '#markup' => $this->t("The imported @incoming with the ID '%id' tries to override a local %local with the same ID.", $arguments),
+          ];
+        }
+
+        $needs_save = $this->updateAdmsFields($local_entity, $incoming_entity);
+
+        // Cleanup the entity from the 'staging' graph.
+        $this->getRdfStorage()->deleteFromGraph($id, 'staging');
+      }
+      // No local entity. Copy the incoming entity as a published entity.
+      else {
+        $local_entity = (clone $incoming_entity)
+          ->enforceIsNew()
+          ->setOwnerId($this->currentUser->id());
+
+        // Determine the state field for this bundle, if any.
+        $state_field_name = NULL;
+        $state_field_map = $this->entityFieldManager->getFieldMapByFieldType('state');
+        if (!empty($state_field_map['rdf_entity'])) {
+          foreach ($state_field_map['rdf_entity'] as $field_name => $field_info) {
+            if (isset($field_info['bundles'][$bundle])) {
+              $state_field_name = $field_name;
+              break;
+            }
+          }
+        }
+
+        // There are also entities without a state field.
+        if ($state_field_name) {
+          $local_entity->set($state_field_name, 'validated');
+        }
+        $local_entity->graph->value = 'default';
+
+        // A new entity needs to be saved.
+        $needs_save = TRUE;
+
+        // Delete the incoming entity from the staging graph.
+        $incoming_entity->skip_notification = TRUE;
+        $incoming_entity->delete();
+      }
+
+      if ($needs_save) {
+        $local_entity->skip_notification = TRUE;
+        $local_entity->save();
```

Comment:
Are we sure that we want to save the entities here? We have the ticket about filtering the solutions to import and we should at least add a `@todo` that the provenance entities should be handled along.

Though I still consider whether this should be handled in the form submit of the entities to import filtering.

Reply:
The provenance entities will be created/updated one step before this step, in user filtering step. At that point the entities are already in the `staging` graph. What are we doing here is only to move them from `staging` to `default` for new entities and copy over for existing entities. I think this was what has been discussed yesterday with Sander.